smb/smb.go

```diff
@@ -754,6 +832,594 @@ type QueryDirectoryRes struct {
 	Buffer             []byte
 }
 
+type QueryInfoReq struct {
+	Header
+	StructureSize         uint16 // Must always be 41 regardless of Buffer size
+	InfoType              byte
+	FileInfoClass         byte
+	OutputBufferLength    uint32
+	InputBufferOffset     uint16
+	Reserved              uint16
+	InputBufferLength     uint32
+	AdditionalInformation uint32
+	Flags                 uint32
+	FileId                []byte
+	Buffer                []byte
+}
+
+type QueryInfoRes struct {
+	Header
+	StructureSize      uint16 // Must always be 9
+	OutputBufferOffset uint16
+	OutputBufferLength uint32
+	Buffer             []byte
+}
+
+type SecurityDescriptor struct {
+	Revision    uint16
+	Control     uint16
+	OffsetOwner uint32
+	OffsetGroup uint32
+	OffsetSacl  uint32 // From beginning of struct?
+	OffsetDacl  uint32 // From beginning of struct?
+	OwnerSid    *SID
+	GroupSid    *SID
+	Sacl        *PACL
+	Dacl        *PACL
+}
+
+type PACL struct {
+	AclRevision uint16
+	AclSize     uint16
+	AceCount    uint32
+	ACLS        []ACE
+}
+
+// MS-DTYP Section 2.4.4.1 ACE_HEADER
+type ACEHeader struct {
+	Type  byte
+	Flags byte
+	Size  uint16 //Includes header size?
+}
+
+type ACE struct {
+	Header ACEHeader
+	Mask   uint32
+	Sid    SID //Must be multiple of 4
+}
+
+type SID struct {
+	Revision       byte
+	NumAuth        byte
+	Authority      []byte
+	SubAuthorities []uint32
+}
+
+func (self *QueryInfoReq) MarshalBinary(meta *encoder.Metadata) (ret []byte, err error) {
+	log.Debugln("In MarshalBinary for QueryInfoReq")
+	buf := make([]byte, 0, 40+len(self.Buffer))
+
+	hBuf, err := encoder.Marshal(self.Header)
+	if err != nil {
+		log.Debugln(err)
+		return nil, err
+	}
+	buf = append(buf, hBuf...)
+	// StructureSize
+	buf = binary.LittleEndian.AppendUint16(buf, self.StructureSize)
+	// Info Type
+	buf = append(buf, self.InfoType)
+	// FileInfoClass
+	buf = append(buf, self.FileInfoClass)
+	// OutputBufferLength
+	buf = binary.LittleEndian.AppendUint32(buf, self.OutputBufferLength)
+	// InputBufferOffset
+	buf = binary.LittleEndian.AppendUint16(buf, uint16(24+len(self.FileId)))
```

Comment:
Three things:
- self.FileId is always 16 bytes length.
- "The offset, in bytes, from the beginning of the SMB2 header to the input buffer", so not from beginning of the QueryInfoReq struct. Add size of SMB2 header to calculation.
- This field should be set to 0 if the input buffer is empty.